api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java

```diff
@@ -0,0 +1,209 @@
+import org.springframework.stereotype.Component;
+
+/**
+ * Data integrity rules for the TB page which are:
+ * <ol>
+ *  <li>Patients with missing TB identifiers</li>
+ *  <li>Patients with similar TB identifiers</li>
+ *  <li>Patients with no final outcome after 9 months at start of treatment </li>
+ * </ol>
+ */
+@Component
+public class InvalidTBEncounters extends BasePatientRuleDefinition {
+
+
+
+	@Override
+	public List<RuleResult<Patient>> evaluate() {
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		
+		ruleResults.addAll(patientsWithMissingTBNumbers(Concepts.UNIT_TB_NUMBER, "Unit TB Number"));
+		ruleResults.addAll(patientsWithMissingTBNumbers(Concepts.HSD_TB_NUMBER, "HSD TB Number"));
+		ruleResults.addAll(patientsWithMissingTBNumbers(Concepts.DISTRICT_TB_NUMBER, "District TB Number"));
+		ruleResults.addAll(singlePatientWithDuplicateTBNumberAcrossMultipleTreatmentPrograms(Concepts.DISTRICT_TB_NUMBER, "District TB Number"));
+		ruleResults.addAll(singlePatientWithDuplicateTBNumberAcrossMultipleTreatmentPrograms(Concepts.UNIT_TB_NUMBER, "Unit TB Number"));
+		ruleResults.addAll(singlePatientWithDuplicateTBNumberAcrossMultipleTreatmentPrograms(Concepts.HSD_TB_NUMBER, "HSD TB Number"));
+		ruleResults.addAll(multiplePatientsWithTheSameTBIdentifiers(Concepts.DISTRICT_TB_NUMBER, "District TB Number"));
+		ruleResults.addAll(multiplePatientsWithTheSameTBIdentifiers(Concepts.UNIT_TB_NUMBER, "Unit TB Number"));
+		ruleResults.addAll(multiplePatientsWithTheSameTBIdentifiers(Concepts.HSD_TB_NUMBER, "HSD TB Number"));
+		ruleResults.addAll(patientsWithNoFinalOutcomeNineMonthsAfterStartOfTreatment(DateTime.now()));
+		return ruleResults;
+	}
+
+	/**
+	 * Patients with similar TB identifiers: duplicated across multiple patients
+	 * @return List<RuleResult<Patient>>
+	 */
+	public List<RuleResult<Patient>> multiplePatientsWithTheSameTBIdentifiers(String identifierConceptUuid, String identifierTitle) {
+		log.info("Executing rule to find Patients with similar TB identifiers: duplicated across multiple patients");
+		String queryString = "SELECT o FROM Obs o WHERE o.voided = false AND o.encounter.patient.dead = 0 AND o.encounter.voided = 0 AND o.concept.uuid = :identifierConceptUuid AND o.valueText IN " 
+				+ " (SELECT ob.valueText FROM Obs ob WHERE ob.concept.uuid = :identifierConceptUuid AND ob.voided = 0 GROUP BY ob.concept.conceptId, ob.valueText HAVING COUNT(ob.valueText) > 1)"; 
+		Query query = getSession().createQuery(queryString);
+		query.setParameter("identifierConceptUuid", identifierConceptUuid);
+		
+		List<Obs> obsList = query.list();
+		Set<Patient> uniquePatientList = new HashSet<Patient>();
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Obs obs : obsList) {
+			Patient patient = obs.getEncounter().getPatient();
+			
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
```

Comment:
@moshonk The action url should go to the form with the duplicated TB number 

Also the message shown is "The District TB Number is used by another patient" - without specifying the TB number